api/src/main/java/org/apache/iceberg/catalog/SupportsCatalogTransactions.java

```diff
@@ -0,0 +1,137 @@
+     * of throughput and resource consumption.
+     */
+    PESSIMISTIC,
+
+    /**
+     * Run the transaction with an optimistic behavior. This assumes that most operations will not
+     * conflict. It typically increases concurrency and throughput while reducing resource
+     * consumption.
+     */
+    OPTIMISTIC;
+  }
+
+  /**
+   * Get the list of {@link LockingMode}s this Catalog supports.
+   * @return A set of locking modes supported.
+   */
+  default Set<LockingMode> lockingModes() {
+    return Collections.emptySet();
+  }
+
+  /**
+   * Get the list of {@link IsolationLevel}s this Catalog supports.
+   * @return A set of locking modes supported.
+   */
+  default Set<IsolationLevel> isolationLevels() {
+    return Collections.emptySet();
+  }
+
+  /**
+   * Start a new transaction and return a {@link TransactionalCatalog} that supports
+   * cross table operations. If a {@link Catalog} supports {@link OPTIMISTIC}
+   * locking mode, this will use that mode. If it only supports {@link PESSIMISTIC},
+   * it will use that mode with default values for lock times.
+   *
+   * @param isolationLevel The {@link IsolationLevel} to use for the transaction.
+   * @return A Catalog with an open transaction.
+   * @throws IllegalArgumentException If the IsolationLevel is not supported.
+   */
+  TransactionalCatalog createTransaction(IsolationLevel isolationLevel);
+
+  /**
+   * Start a new transaction that holds locks for the life of the transaction.
+   *
+   * <p>If supported, this type of transaction allows a much higher likelihood of
+   * transaction completion but can also result in more failures, lower concurrency
+   * and deadlocks.
+   *
+   * @param isolationLevel The {@link IsolationLevel} to use for the transaction.
+   * @param lockWaitInMillis How long to wait in milliseconds when grabbing a pessimistic lock.
+   * @param maxTimeMillis The maximum amount of time the transaction is allowed to be open.
+   * @param tables An optional list of tables that should be locked immediately for this
+   *        transaction. Note that this can include tables that exist and tables that do not
+   *        exist as this transaction may be adding new tables and want to grab pessimistic locks
+   *        for those operations.
+   *
+   * @return A Catalog with an open transaction.
+   *
+   * @throws IllegalArgumentException If the IsolationLevel is not supported. Also throws if
+   *         {@code lockWaitInMillis} or {@code maxTimeMillis} are not within supported limits
+   *         (check specific {@link Catalog} documentation for any limits).
+   * @throws CommitFailedException If the list of {@code tables} cannot be locked.
+   * @throws ValidationException If the set of tables names includes tables names that are not
+   *         valid.
+   */
+  TransactionalCatalog createLockingTransaction(
+      IsolationLevel isolationLevel,
+      long lockWaitInMillis,
+      long maxTimeMillis,
+      TableIdentifier...tables);
```

Comment:
Why supply tables here? Isn't the reason to use a `Catalog` so that any table that is loaded and scanned is validated as part of the transaction?

Reply:
Per the Javadoc, this is an optional list of tables you can grab locks on at the beginning of a transactions. If you provide no tables, then the expectation would be that locks would be acquired as you interacted with tables. This allows you to fail-fast if you can't grab some locks rather than do a bunch of work and then fail to get a last lock later. At least that was the intention with the api. (This is only relevant to the pessimistic api as there is no such thing as acquiring locks in the optimistic api)

Reply:
Sorry I missed that in Javadoc. That sounds reasonable to me.

Reply:
I added a little more doc and renamed the argument tablesHint to help clarify the purpose.